dask/bag/core.py

```diff
@@ -1630,8 +1646,9 @@ def from_delayed(values):
     return Bag(graph, name, len(values))
 
 
-def merge_distinct(seqs):
-    return set().union(*seqs)
+@curry
+def merge_distinct(seqs, key=toolz.identity):
+    return toolz.pipe(seqs, toolz.concat, toolz.curried.unique(key=key), list)
```

Comment:
The use of hardcore `toolz` concerns me here for two reasons:

1.  While concise, I suspect that it is a bit too complex/foreign for some developers who might come by in the future.  If possible I think that it would be good to do this with more traditional code.
2.  I'm concerned that it might be slower.  

    For example I suspect that `toolz.curried.get(key)` is both more confusing and slower to execute than `lambda x: x[key]`

Reply:
2. I think you are right

```python3
In [11]: from toolz.curried import get
    ...: x = {'a': 0}
    ...: f = get('a')
    ...: g = lambda x: x['a']
    ...: %timeit -n 1000 f(x)
    ...: %timeit -n 1000 g(x)
708 ns ± 69.1 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
124 ns ± 0.403 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

Reply:
Do you mean something like:


```python3
@curry
def merge_distinct(seqs, key=None):
    if key is None:
        return toolz.pipe(seq, toolz.concat, list)
    else:
        if not callable(key):
            key = lambda x: x[key]
        return toolz.pipe(seq, toolz.concat, toolz.curried.unique(key=key), list)
```

Or should I remove `@curry` and `toolz.curried.*` as a whole?

Reply:
Well, in the first case I would probably just call `return list(concat(seq))`, which seems simpler and doesn't require readers to understand pipe.  

Pipe and curried can sometimes be really critical.  I don't think that this is one of those times though.